Migration guide for window singleton deprecation (#8283)
flutter/website: src/release/breaking-changes/window-singleton.md

---
title: The window singleton is deprecated
description: In preparation for supporting multiple views and multiple windows the window singleton has been deprecated.
---

> ## Summary

Reviewer: This paragraph feels like more suitable for the "Context" section. This section, to my understanding, should summarize the changes and symptoms. For example:

A number of APIs that assume the app only has one view have been removed. Including:

* `GestureBinding.instance.window`.
* <etc>

Accessing these APIs might result in compiling errors such as:

```
Error: Member not found: 'window'.
```

Author: Yeah, some of this should move to the context section. I'll do that and provide a more high-level summary of the change itself here.

> Originally, Flutter assummed that an application would only consist of a single

Reviewer 2: ```suggestion
Originally, Flutter assumed that an application would only consist of a single
```

view (the `window`) into which content can be drawn. In preparation for
supporting multiple views and multiples windows, this single view assumtion and
the APIs that exposed it have been deprecated. Applications and libraries that
relied on these APIs need to migrate to new multi-view compatible APIs as
outlined in this migration guide.

## Context

> In a multi-view world APIs that just assume a single window do not make sense

Reviewer 2: ```suggestion
In a multi-view world, APIs that assume a single window do not make sense
```

anymore. Instead, applications and libraries must chose the specific
view they want to operate on. 

## Description of change

The APIs that have been deprecated as part of this change are:

> * the global `window` property exposed by `dart:ui`

Reviewer 2: Should these be capitalized?

```suggestion
* The global `window` property exposed by `dart:ui`
```

* the `window` property on the `BaseBinding` class, 
  which is usually accessed via
  * `GestureBinding.instance.window`
  * `SchedulerBinding.instance.window`
  * `ServicesBinding.instance.window`
  * `PaintingBinding.instance.window`
  * `SemanticsBinding.instance.window`
  * `RendererBinding.instance.window`
  * `WidgetsBinding.instance.window`
  * `WidgetTester.binding.window`
* the `SingletonFlutterView` class from `dart:ui`
* TODO: mention any test related classes/properties (e.g. TestWindow)

> The following options exist to migrate application and library code that relies

Reviewer: This section should probably go to the "Migration guide" section.

Author: My intention was to keep the migration guide section really short and focused on the actual code changes you need to make. This section on the other hand provides a little more detail on the new API.

on these deprecated APIs:

> If a `BuildContext` is available, consider looking up the current `FlutterView`

Reviewer 2: Is second person acceptable in migration docs? If so, consider something like:

```suggestion
If a `BuildContext` is available, you can use `View.of` to get the current `FlutterView`.
```

Author: I think migration guides should be in passive voice like our API docs?

via `View.of`. This returns the `FlutterView` into
which the `build` method associated with the given context is drawing the
widgets it builds. The `FlutterView` provides access to the same functionality
that was previously available on the deprecated `SingletonFlutterView` class
returned by the deprecated `window` properties mentioned above. However, some
of the platform-specific functionality has moved to the `PlatformDispatcher`,
which can be accessed from the `FlutterView` returned by `View.of` via 

> `FlutterView.platformDispatcher`. Using `View.of` is the prefered way of

Reviewer 2: ```suggestion
`FlutterView.platformDispatcher`. Using `View.of` is the preferred way of
```

  addTeardown(tester.platformDispatcher.clearAccessibilityFeaturesTestValue);
  addTeardown(tester.platformDispatcher.clearAlwaysUse24HourFormatTestValue);
  addTeardown(tester.platformDispatcher.clearBrieflyShowPasswordTestValue);
  addTeardown(tester.platformDispatcher.clearDefaultRouteNameTestValue);
  addTeardown(tester.platformDispatcher.clearInitialLifecycleStateTestValue);
  addTeardown(tester.platformDispatcher.clearLocalesTestValue);
  addTeardown(tester.platformDispatcher.clearLocaleTestValue);
  addTeardown(tester.platformDispatcher.clearNativeSpellCheckServiceDefinedTestValue);
  addTeardown(tester.platformDispatcher.clearPlatformBrightnessTestValue);
  addTeardown(tester.platformDispatcher.clearSemanticsEnabledTestValue);
  addTeardown(tester.platformDispatcher.clearTextScaleFactorTestValue);
});
```

##### Resetting all properties at once

Code before migration:

```dart
testWidgets('test name', (WidgetTester tester) async {
  addTeardown(tester.binding.window.clearAllTestValues);
});
```

Code after migration:

```dart
testWidgets('test name', (WidgetTester tester) async {
  addTeardown(tester.platformDispatcher.clearAllTestValues);
});
```

## Timeline

> Landed in version: TODO<br>

Author: Reminder to myself to fill this in before submitting this guide. The change is expected to land sometime today or tomorrow.

  * `WidgetsBinding.instance.window`, or
  * `WidgetTester.binding.window`.
* The `SingletonFlutterView` class from `dart:ui`.
* `TestWindow` from `flutter_test`, its constructors,
  and all of its properties and methods.

> * The `WidgetTester.binding.window` accessor.

Author: This one is already listed on line 36.

Reviewer 3: Good catch, removed!

`PlatformDispatcher.instance` property. This ensures that the functionality
of the `PlatformDispatcher` can be properly mocked out in tests.

### Testing

For tests that accessed the `WidgetTester.binding.window` property to change
window properties for testing, the following migrations are available:

In tests written with `testWidgets`, two new properties have been added that

> together replace the functionality of `TestWindow`

Author: nit: this sentence is missing punctuation at the end.

Reviewer 3: Another good catch, added!
